File: app/src/main/java/com/elitec/alejotaller/feature/product/domain/entity/StockMovement.kt
package com.elitec.alejotaller.feature.product.domain.entity

/**
 * Movimiento de inventario (colección Appwrite `stock_movements`).
 * quantity siempre > 0; el sentido lo define [type].
 */
enum class StockMovementType {
    ENTRADA,
    SALIDA_VENTA,
    AJUSTE,
    DEVOLUCION
}

data class StockMovement(
    val id: String,
    val productId: String,
    val type: StockMovementType,
    val quantity: Int,
    val balanceAfter: Int,
    val reason: String,
    val userId: String,
    val saleId: String? = null,
    /** ISO-8601 o epoch millis según capa de persistencia. */
    val createdAtIso: String
) {
    init {
        require(id.isNotBlank()) { "StockMovement id cannot be blank" }
        require(productId.isNotBlank()) { "productId cannot be blank" }
        require(quantity > 0) { "quantity must be positive" }
        require(balanceAfter >= 0) { "balanceAfter cannot be negative" }
        require(reason.isNotBlank()) { "reason is required" }
        require(userId.isNotBlank()) { "userId is required" }
    }
}

File: web/src/core/feature/product/domain/entity/StockMovement.ts
export type StockMovementType =
    | "entrada"
    | "salida_venta"
    | "ajuste"
    | "devolucion";

export interface StockMovement {
    id: string;
    productId: string;
    type: StockMovementType;
    /** Siempre positivo; el sentido lo define `type`. */
    quantity: number;
    /** existence del producto después del movimiento. */
    balanceAfter: number;
    reason: string;
    userId: string;
    saleId?: string | null;
    createdAt: string; // ISO-8601
}

export function createStockMovement(input: StockMovement): StockMovement {
    if (!input.id?.trim()) throw new Error("StockMovement id cannot be blank");
    if (!input.productId?.trim()) throw new Error("productId cannot be blank");
    if (!input.quantity || input.quantity <= 0) {
        throw new Error("quantity must be positive");
    }
    if (input.balanceAfter == null || input.balanceAfter < 0) {
        throw new Error("balanceAfter cannot be negative");
    }
    if (!input.reason?.trim()) throw new Error("reason is required");
    if (!input.userId?.trim()) throw new Error("userId is required");

    return {
        ...input,
        quantity: Math.floor(input.quantity),
        balanceAfter: Math.floor(input.balanceAfter)
    };
}
